### ISSUE TITLE
Niche issue when streaming RA using some shader presets

### ISSUE THREAD

**Issue by klepp0906** [libretro/RetroArch#9340]
# First and foremost consider this:
- Only RetroArch bugs should be filed here. Not core bugs or game bugs
- This is not a forum or a help section, this is strictly developer oriented

This is probably exceptionally low priority and something i'll have to live with.   Still - I wanted to make sure it was put out there.   As of a few versions back, I am no longer able to stream RA properly, at least in the means I have for a long time now.  Unfortunately I'm a serial updater, and ive had a good week off of playing, so i cant say exactly what or when caused the regression.

The short of it is this.  

I have content directory shader presets set up.  In this case, for beetle psx hw.  I load a game from said playlist, config override kicks in and positions window properly/places within overlay, and shader preset kicks in which ties into the configured sizing present in the config.  

this has stopped working if I am streaming and if im using my go-to shader that I have set up (ctr guest dr venom.)   I stream via steam BPM via steam link app and if I load retroarch and load any playlist which uses that shader, I get a black screen with audio.  This also happens with several other higher end crt shaders.   

If i come back to my pc, the image is on the screen, but is without color.  If i remove the shader preset, all is fine.  If i load a basic shader such as Geom's as a preset, that works fine as well.

One other oddity.   If i load a game via command line/shortcut direct from steam, it works fine.   ONLY if i load it via retroarch itself does it fail.

Im not sure if something a few versions back changed the priority or order these things were drawn or loaded - but definitely puts a damper on my vastly preferred means of playing.  

at current my options are command line shortcuts/launching of any game i want to play that uses that shader.  The removal of those shader presets entirely from ~30 playlists and go with something basic like geoms, and resize all my config files to accomodate.  Or, play exclusively from my pc.

This could very well stem elsewhere, however I did try several different gpu drivers just in case.   Also tried removing curve to eliminate that as being a potential variable as i know royale fails when curve is added to the picture.

The strange part is,  if i remove the preset, and load it manually after the game is loaded - it works just fine as well.  its purely giving me hell when loaded automatically via content directory preset.

### Expected behavior

The game to display alongside overlay AND shader, instead of overlay + black screen + audio.

### Actual behavior

See above. 

### Steps to reproduce the bug

1. Launch retroarch via (presumably) a steam link - or in my case the steam link app. 
2. Load a game from a playlist that uses crt-guest-dr-venom
3. note everything loads, (can see overlay, audio is playing) the gameplay area is black.

### Bisect Results

I cant say exactly, as ive been not playing (but updating) for the past week or so.  But its very recent. 

### Version/Commit
You can find this information under Information/System Information

- RetroArch: 1.78 9a36851beb

### Environment information

Windows 10

**Comment by LazyBumHorse** [comment]
Hm... I don't get the full picture of the issue yet, but let's start with this:

> The strange part is, if i remove the preset, and load it manually after the game is loaded - it works just fine as well. its purely giving me hell when loaded automatically via content directory preset.

Can you try the following:

1. In RetroArch go to Settings -> Drivers -> Video and set it to "glcore"
1. Restart RetroArch
1. Load a game and create a content preset with `crt-guest-dr-venom.slangp`

After this, does the issue still happen if you load a game from that same content directory?
If it works with glcore, there might be something up with the "gl" video driver.
If it doesn't work, can you change the video driver back to "gl" and test the [nightly version from 2019-08-16](http://buildbot.libretro.com/nightly/windows/x86_64/2019-08-16_RetroArch.7z)?
If that works, can you then test if the [next version, from 2019-08-17](http://buildbot.libretro.com/nightly/windows/x86_64/2019-08-17_RetroArch.7z) is broken?

Somewhere has to be the point where it first broke and if we can pinpoint it, it'll be much easier to find the issue.

I also got some question just to understand the issue a bit better:

>  I load a game from said playlist, config override kicks in and positions window properly/places within overlay

The overlay is from the streaming software, is that right?

> If i come back to my pc, the image is on the screen, but is without color.

By "the image", do you mean you can see both the overlay + the game like it should be - just both in black and white?

**Comment by klepp0906** [comment]
hi and thanks for replying.   Didnt expect anyone to bother.   Retroarch is available on so many platforms I dont imagine "too" many people stream.   Still until I can get a very hefty NAS coupled with something powerful enough to emulate the newer stuff @ 4k, streaming is my go-to and has been for a long time.

fwiw am already using glcore as i believe regular ole GL isnt compatible with slang shaders which motivated my switch to glcore from gl awhile back.

i will try some even older versions as soon as i can.  I went back a couple days with no luck, not so far as the 16 though.  I purely dragged and dropped the exe from the buildbot over, hope thats adequate? 

The overlay is the overlay i set within retroarch.  A border as many also like to call them.  I use console specific ones for each playlist.  The "border" always shows up, as does the popup that says "configuration override loaded" but at that point - i get just a black screen where the game "should" be with the audio playing.

If i leave my tv at that point, and return to my pc, the overlay is there properly in color.  the game isnt a black screen, its working - but is in black and white, not color.   Weird right?

**Comment by klepp0906** [comment]
if you need me to clarify anything further - please.

**Comment by klepp0906** [comment]
I did try the version from the 16th.  It does in fact work properly.   So some point between then and now it stopped functioning properly.

**Comment by klepp0906** [comment]
figured id might as well press on to find out where exactly things went awry - make your job easier.

fortunately for me - only took 1 more day.

it broke on the build from the 17th.

**Comment by twinaphex** [comment]
So on the 17th, I can see that @LazyBumHorse's PR got merged. Is that the likely cause?

@klepp0906 would you be able to bisect further? We have an important release coming up over the weekend, would be nice to get this issue sorted before then.

**Comment by klepp0906** [comment]
uh, of course id be "willing" to bisect further.  But unsure how.   If you want to hand hold and you have releases between the 16th and the 17th, certainly.

**Comment by twinaphex** [comment]
Kinda knee deep in preparing the release unfortunately.

**Comment by klepp0906** [comment]
totally understandable.  Unfortunately the whole "pulling" specific changes between 16th and 17th is a smidgen beyond my understanding.   I assume thats what you were referring to when you said "bisect further" yea?

**Comment by twinaphex** [comment]
yeah you'd just do a git checkout at each commit hash, then recompile and check out the compiled binary.

**Comment by klepp0906** [comment]
i dont even have git installed.  I do have the github desktop client which I "think" can do that - but it would be my first time even opening it lol.

I can take a look, but the likelihood of it bearing fruit is substantially low.   Still - I can try to figure it out :P

**Comment by klepp0906** [comment]
first one i tested `719555bae156f24201026537f2e710f148a63c93` was bad.  
I believe thats the first from that day.

**Comment by LazyBumHorse** [comment]
You did it!
So it _is_ my commit because the nightly from 2019-08-16 is just 1 commit before that.
That's what I feared.

I looked at the changes and I'm not yet sure how this could break your use case.

It seems the only possible changes that could affect this was in `menu_shader_manager_init()` and even there all that's effectively different is a call to `video_shader_driver_get_current_shader()`... it's just weird.

So I made a custom debug branch here: https://github.com/LazyBumHorse/RetroArch/tree/debug_9340
Can you pull and build this?

It should output a log with lots of "[DEBUG]" messages if you enable Log Verbosity and Log To File under Settings -> Logging (make sure to restart RetroArch after that).
So if you go through the steps to make it fail, it could tell us what's wrong.

You can upload the log on pastebin or similar.

**Comment by klepp0906** [comment]
i can do that, but big problem is this.

I tried enabling logging previously to get some information as to why this happened, something more for you guys to go on.

It doesnt seem to happen when logging is enabled lol.   I assume it has something to do with the fact that when logging is enabled, retroarch opens simultaneously more or less with a cmd window.   Just briefly it pops up before RA draws, you see it when you start the stream.   

somehow that circumvented the issue.

either way, I can certainly build that and try it and see what the log says.

never built a debug build but i think the directions are in the docs/shouldnt be too much different..

will get on to that sometime this morning.   (gotta feed the tyrants and head to gym rq, then ill be on it)

**Comment by LazyBumHorse** [comment]
> It doesnt seem to happen when logging is enabled lol. I assume it has something to do with the fact that when logging is enabled, retroarch opens simultaneously more or less with a cmd window. Just briefly it pops up before RA draws, you see it when you start the stream.

Hm... maybe the streaming software grabs the window a bit later and doesn't cause issues that way.
Does the streaming software have some kind of "hook" delay you can increase?
As far as I understand, such software puts on a kind of overlay like the Steam overlay that can grab the screen/window's content but it's not a failure free process, especially if software just starts up - and when you start a game from RetroArch, it does do a video reinitialization which is probably what screws things up.

I believe I ran into similar issues with RetroArch when I was using MSI Afterburner or rather RivaTuner, I ended up blacklisting RetroArch because of it.

But let's still try this:
I think you can actually disable Log Verbosity, just tested it, the [DEBUG] messages are still written and the window is then gone.

> never built a debug build but i think the directions are in the docs/shouldnt be too much different..

You don't need to do anything different, it's actually better to just do a normal/release build, so nothing really changes except that some log messages are written.

**Comment by klepp0906** [comment]
ok so just clone it, build it normal, and enable logging?

**Comment by LazyBumHorse** [comment]
yep, logging to file (which will write to logs/retroarch.log)

**Comment by LazyBumHorse** [comment]
Oh wait, I'm an idiot, it doesn't actually write the file without verbosity...

**Comment by klepp0906** [comment]
lol well its building now either way ;p

**Comment by LazyBumHorse** [comment]
I just pushed another commit to the branch which disables the console.
So log verbosity + log to file writes the file without any extra window.

**Comment by klepp0906** [comment]
k, will pull and rebuild

**Comment by klepp0906** [comment]
![image](https://user-images.githubusercontent.com/10392920/63589469-402d4700-c577-11e9-9a91-8658e3aa2a44.png)

still popping up that which is causing it to "work"

**Comment by klepp0906** [comment]
heres the log, though probably not helpful since it "worked"
https://pastebin.com/MKj8CPeL

**Comment by klepp0906** [comment]
def still fails once logging is disabled.  returned to pc and RA is playing in black and white :p

**Comment by LazyBumHorse** [comment]
Hm... that doesn't have any "[DEBUG]" messages, I think you built the "master" branch, you need to switch/checkout to the "debug_9340" branch.
In the command line that's `git checkout debug_9340`.

**Comment by LazyBumHorse** [comment]
Actually, I think you need to fetch that first also, hold on, I'm not exactly well versed with git myself

**Comment by klepp0906** [comment]
ok, probably what happened then.

I cloned straight from the git repo link you gave me.

so let me get this right.   Clone from that repo `https://github.com/LazyBumHorse/RetroArch/tree/debug_9340`

then do a `git checkout debug_9340`

then ./configure
make clean
make -j4

and im set?

**Comment by klepp0906** [comment]
hah, you posted right before me.  Oh boy.  well im gonna be bailing relatively soon for a couple of hours so just trying to hammer out what i can. 

If i disappear ill be back on it asap

**Comment by LazyBumHorse** [comment]
If you do a full clone of https://github.com/LazyBumHorse/RetroArch.git and then `git checkout debug_9340` that should definitely work.

I'm sure there's an easier way to update just that branch, maybe `git checkout -B debug_9340 origin/debug_9340` followed by a `git pull`.

**Comment by klepp0906** [comment]
built it doing what i said, and the resultant file size is different, so we're gonna assume it worked.

also i was doing strip -s retroarch out of habit, but i recall someone saying that strips debug stuff maybe?

either way, gonna give this build a go RQ

**Comment by klepp0906** [comment]
looks like the build just has core logging level, no more frontend logging level toggle.   so i have to presume is the right build

**Comment by klepp0906** [comment]
ok, no window popped up to mess things up, and as such - it didnt work properly. 

got the log.  I hope its helpful.

https://pastebin.com/kRdU5DpL

**Comment by LazyBumHorse** [comment]
Thanks, this will take some time to parse

**Comment by klepp0906** [comment]
also thank you for looking into this.  All you guys.  I totally thought with all the stuff RA has going on simultaneously this wouldnt be deemed important enough to fix.   At least not right away. 

Course im biased as all hell, but from my perspective - RA without shaders ewww.   And once you can RA from your couch/recliner/bed - you never go back :P

id love to simply have a capable HTPC at all my tv's but yea lol

**Comment by LazyBumHorse** [comment]
So the log looks completely fine, in fact, apart from the debug messages, it is identical to the "working" log above, except this for this line
> [INFO] Environ SET_MESSAGE: Slow CD image read detected -- consider using async or precache CD Access Method

which may just be of random disk spin-up times.

The log did give a tiny hint though, there's a `command_event(CMD_EVENT_SHADER_PRESET_LOADED, NULL)` call in `menu_shader_manager_init()` where one previously wouldn't happen, so I'm thinking we should try different versions of it and see if that changes anything.

Let's see what happens if we kill the `menu_shader_manager` completely.

I updated the branch, so you can do a
`git fetch origin`
`git checkout -B debug_9340 origin/debug_9340`
and rebuild.


> Course im biased as all hell, but from my perspective - RA without shaders ewww

My exact stance, it's one of the best features about RA because you get the whole shader range for almost all possible emulators.
It's why I got into contributing to RetroArch.

**Comment by klepp0906** [comment]
okie dokie, im on it boss.  report back as soon as i can.   (these things take 10x as long, half cause i dont know what im doing, half cause my toddlers perpetually attack me)

**Comment by klepp0906** [comment]
noooo bueno on that one.  same behavior.

here is the log https://pastebin.com/YcwPCK6X

**Comment by LazyBumHorse** [comment]
That's really surprising!

I guess I need to go through all changes again, but I currently have no idea what's going on.

Is it possible it's not this commit after all and it was a random fluke? Maybe build related?

Can you try to build this commit here: https://github.com/libretro/RetroArch/commit/77d2ce95f1c30c2e5b0d4a8ba97f2c4391b7cc99
So you go into your RetroArch repo (shouldn't matter if it's mine or the official one) and checkout to the commit `77d2ce95f1c30c2e5b0d4a8ba97f2c4391b7cc99` which is named "Update Japanese translation", build and test it.

**Comment by klepp0906** [comment]
i dont think its possible, as i switched back and forth between several builds on the buildbot initially.   I dont build my own builds if i can at all avoid it.

will build that build now and report back with a log asap.

**Comment by klepp0906** [comment]
ok, so that one works.

i imagine, what your going to tell me is that commit you just gave me, is one that comes after yours, and since it works.  It is in fact not that initial commit that breaks it, and it was a flub on my end.

**Comment by LazyBumHorse** [comment]
No, it's the one right before it, so I guess we have revert my changes part by part and see when it's starting to work again.

Another question: Do any of your configs or overrides use the "video_shader" setting?

**Comment by klepp0906** [comment]
well, at least i dont have to feel like a pleb then :)

what is the video_shader settings?  aka what does it control?

all my config overrides are the same in content, heres an example.

this is Sony - PlayStation.cfg

aspect_ratio_index = "23"
custom_viewport_height = "1988"
custom_viewport_width = "2672"
custom_viewport_x = "584"
custom_viewport_y = "81"
input_overlay = ":\overlays\custom\OrionsangelRetroarchOverlays\originals\playstation.cfg"
video_scale_integer = "false"

as for the Shaders
its just 

Sony - PlayStation.slang which loads the crt guest dr venom shader preset automatically when anything is loaded from the playstation directory.

as mentioned earlier, if i change it from guest dr venom to something simple like vanilla geoms, it works.  However other advanced shaders like lottes etc do not.

unsure if its something that resides in those shaders thats contributing, or the fact theyre more complex and load slower relative to geoms.

thats why we rely on the big guns - aka you guys :)

**Comment by LazyBumHorse** [comment]
Okay, let's do some more testing, I actually have an idea, but we should test this one first.

So if you go to where you cloned my repo, you can type in
`git fetch origin`
`git checkout -B debug2 origin/debug2`
build and test.

Don't need a log this time.

**Comment by LazyBumHorse** [comment]
Btw, the video_shader setting was a way to set a shader and it was removed in my commit because of bad behavior (and because presets are generally better).

**Comment by klepp0906** [comment]
ah, well fortunately sounds like it didnt/doesnt apply to me/my use.

got the new one building now, will letcha know stat.

**Comment by klepp0906** [comment]
no bueno on that build.  still broken.

**Comment by LazyBumHorse** [comment]
Okay, now for the weird test.
fetch, checkout, build, test, you know how it's done. :smiley:

**Comment by klepp0906** [comment]
same info as above? 

git fetch origin
git checkout -B debug2 origin/debug2  ?

**Comment by LazyBumHorse** [comment]
yep!

**Comment by klepp0906** [comment]
i have a question, after fetching and checking out - do i need to redo the ./configure step before each build at that point? 

never closed msys nor left the directory if that makes a difference

**Comment by LazyBumHorse** [comment]
I'm not sure when exactly what the conditions are, but `./configure` is rarely needed, you'll know if you need it, when it won't `make` at all.
I think it's needed when the makefile changes, but in this specific case I'm only changing source files, so if you called it once, you're good.

**Comment by LazyBumHorse** [comment]
Personally I have a `./conf` script that does `./configure && make clean` that I occasionally call and then a `./c` script that does `make DEBUG=1 -j15 >/dev/null && ./retroarch` (you should definitely use `-j<thread count>`, it'll lower build times significantly)

**Comment by klepp0906** [comment]
yea i use -j4  now i know what that means though :)  can probably get away with -j8 so theres that at least for future use. 

anyhow - that build, no dice :(

**Comment by LazyBumHorse** [comment]
> anyhow - that build, no dice :(

Damn, thought for sure that would do something, not much left, really.

**Comment by LazyBumHorse** [comment]
And once more - this might be too much this time.
Same branch as before.

**Comment by klepp0906** [comment]
on it.

**Comment by klepp0906** [comment]
super disappointing, but no good.  the git version should be ce985dff06 yes?

**Comment by LazyBumHorse** [comment]
Yes that's right.
I pushed another version, let's see what's gonna happen.

**Comment by klepp0906** [comment]
sorry was feeding the terrorists.   back on it.

**Comment by klepp0906** [comment]
hallelujah.  that ones the ticket :)

**Comment by klepp0906** [comment]
do tell - what was it?  (in terms a nobody can understand!)

**Comment by LazyBumHorse** [comment]
In terms that anybody can understand: This makes literally 0 sense...

**Comment by klepp0906** [comment]
LOL.  well you had to change what you changed for a reason right?

**Comment by LazyBumHorse** [comment]
Can you rebuild once more and send me the log?

**Comment by klepp0906** [comment]
you betcha.  on it

**Comment by klepp0906** [comment]
full disclosure - that one spit out the same window i posted above, hopefully it doesnt affect the log as far as it having forced it to "work" in the past

https://pastebin.com/W9CcLhAh

**Comment by LazyBumHorse** [comment]
Can you send the content of
`C:\msys64\home\klepp0906\packaged\RetroArch\shaders\retroarch.slangp`
and
`C:\msys64\home\klepp0906\packaged\RetroArch\shaders\presets\Beetle PSX HW\Sony - PlayStation.slangp`?

**Comment by klepp0906** [comment]
https://www.dropbox.com/s/33yvth96jcxtkbv/retroarch.slangp?dl=0

https://www.dropbox.com/s/baq83ct5szztjnu/Sony%20-%20PlayStation.slangp?dl=0

**Comment by LazyBumHorse** [comment]
They are different.

"retroarch.slangp" - what seems to have been used in the working case(?) has:
```
csize = "0.000000"
warpX = "0.000000"
warpY = "0.000000"
masksize = "1.000000"
```
while "Sony - PlayStation.slangp" in the broken case has:
```
csize = "0.030000"
warpX = "0.030000"
warpY = "0.040000"
masksize = "2.000000"
```

Does that mean the preset it loaded was bad all along?

Okay let's try the following:
1. Get the newest nightly
2. Load the shader you want
3. Save as a content preset
4. Try the test again

**Comment by klepp0906** [comment]
that was one of the first things i tried - when the problem cropped up.  also how i found it would load w/ geoms and not with the others.

it has been a few days, a few builds - so ill definitely try again.

unless updating RA has some means of modifying the shader presets....

although now that i look closer at the above, whats different is the curve and corners are applied to one, and not the other.

though the "broken" case is what ive been using all along problem free.   I must have copied my shader presets over when i had it toggled (off).

still, i havent changed the settings in my packaged directory (shader preset settings)  so i dont know how it would be loading one way and then another way.

every time it still had the proper curve applied so even if it was changing it in the file, certainly not visually.

aaaaanyways, will update and try momentarily.

**Comment by LazyBumHorse** [comment]
Hm... no, that's not quite it, it loads retroarch.slangp when going in the menu, but when going ingame it loads Beetle PSX HW\Sony - PlayStation.slangp like it should.

I still don't get why retroarch.slangp is created in the first place though...

**Comment by LazyBumHorse** [comment]
5 times even, did you only start the game or did you interact with the shader menu?

**Comment by klepp0906** [comment]
well. first

the newest nightly - no go.

second, i may have sent you the wrong sony - playstation.slangp from the packaged directory or something.  

I just opened it in notepad and 

![image](https://user-images.githubusercontent.com/10392920/63621112-1a2c9480-c5c1-11e9-88aa-a7cb3d4d7c36.png)

**Comment by klepp0906** [comment]
thats the one that ive used all along.   and the one that works with the working builds, and doesnt work with the non working builds.

here it is from my "live" RA for posterity/comparisons sake.  (thats where that one was copied over from anyhow but /shrug)

https://www.dropbox.com/s/pa2u4f7stsmguea/Sony%20-%20PlayStation%20%282%29.slangp?dl=0

**Comment by klepp0906** [comment]
retroarch.slangp has always been there.  I thought that basically comes with retroarch.  ive never loaded it, used it, modified it.   Dont even know what it is.

**Comment by klepp0906** [comment]
> 5 times even, did you only start the game or did you interact with the shader menu?

i just start the game, and it automatically applies the shader thats set in the Sony - PlayStation.slang

never interacted with the shader menu since looooong ago when i first set it up.   (apart from recently when doing all this testing)

**Comment by LazyBumHorse** [comment]
Sony - PlayStation (2).slangp is the same as the other one.

I think maybe though that there's an issue with loading the shader parameters?
They seem to be treated very differently to any other preset setting and I think I recently discovered a small loading bug that might be related...?

One thing that'd be interesting to test:
Can you load the retroarch.slangp in the newest nightly and save *that* as a content preset?

This will be the last for today, I need to look this over with a clear head, there's too much that doesn't seem to make any sense.

**Comment by klepp0906** [comment]
yea sure.  you want me to save it as a preset and then try to stream?

**Comment by LazyBumHorse** [comment]
Yes, just to see what happens

**Comment by klepp0906** [comment]
same thing.  screen freeze w/ audio.  no bueno.

**Comment by LazyBumHorse** [comment]
Alright, I'll think about all this tomorrow.

**Comment by klepp0906** [comment]
Rest easy :) I'll certainly be around!

for now, back to my 16th backup ;p

**Comment by LazyBumHorse** [comment]
I think I have an idea what's going on.

You can give this a shot:
`git fetch origin`
`git checkout -B debug3 origin/debug3`
build and test.

**Comment by klepp0906** [comment]
on it.

**Comment by klepp0906** [comment]
during building i got this.

not sure if relevant

![image](https://user-images.githubusercontent.com/10392920/63638819-ee5ded00-c65a-11e9-93b5-b6476d2cab19.png)

**Comment by klepp0906** [comment]
it streamed, but its because it didnt load my shader preset for some reason.

So i re applied/configured it, re-saved it, and tried to restart the game.  Still didnt load it.

**Comment by LazyBumHorse** [comment]
Oops, I typo'd it.
You still have your preset at `C:\msys64\home\klepp0906\packaged\RetroArch\shaders\presets\Beetle PSX HW\Sony - PlayStation.slangp`, right?

**Comment by klepp0906** [comment]
yes, correct

**Comment by LazyBumHorse** [comment]
Then you can try the same steps again, hopefully it works this time.

**Comment by klepp0906** [comment]
whelp, that seems to be working :)

**Comment by klepp0906** [comment]
now do you have some laymans explanation? ;p

**Comment by LazyBumHorse** [comment]
Alright!
The thing is really simple.

There's was tiny change in behavior:
Before my commit, shader presets would actually also be loaded in the main menu itself when the `video_shader` setting was set even though you couldn't see them.
After my commit, this does not happen anymore, so why does this matter here?

Because in this case, your `video_shader` setting would load `retroarch.slangp` which is practically identical to your `Sony - PlayStation.slangp`.

So it's loading all the different shader files in the menu and I belive that, the second time it's loaded when you start a game is then much faster, probably because the files are cached in the hard drive.

Basically, the old version would sort of "preload" the shader files which sped things up enough to not mess with the streaming software that's grabbing the window.

**Comment by klepp0906** [comment]
ah, makes sense.  Glad your rested brain prevailed :)

I assume it'll take at least a good solid day for the "fix" to make it into the buildbot right?

**Comment by LazyBumHorse** [comment]
I don't know yet how to tackle this, a "preload" is hard because it requires tons of changes.

A good option would probably be some optional delay, so instead of loading shaders immediately when starting a game, we wait a second or two first.

The _best_ way would be to add a delay to the streaming software!
Are you sure Steam link doesn't have such a thing?

**Comment by klepp0906** [comment]
yea, steam link has very limited options it allows you to mess with unfortunately.   max bandwidth (which i tried the available options) and x265 vs x264 which also didnt help.   Tried hardware and software encoding as well.

What necessitates the change to how it functioned prior to your changes?   Im sure if you did it, theres good reason, just curious.

Cant expect to roll anything back over 1 person, that being said - steam is pretty heavily used so id venture when/if people DO stream, its gonna be steam theyre using.

(this is what i repeat to myself to feel like less of a pain in the ass).

**Comment by LazyBumHorse** [comment]
> What necessitates the change to how it functioned prior to your changes?

The main thing was the removal of the `video_shader` setting, which would always point to the last loaded shader and would load that shader automatically the next time you start RetroArch - even if you never told it to do that. It basically clashes with core/game presets etc.

If you want you can read the full technical description here: https://github.com/libretro/RetroArch/pull/9315

All in all, it's a good thing that shaders are not loaded automatically and especially not loaded in the menu, where it just slows things down without benefit.

Well, except that it acted like a "preload" in your case :smiley: 

> Cant expect to roll anything back over 1 person, that being said - steam is pretty heavily used so id venture when/if people DO stream, its gonna be steam theyre using.

I think OBS is more common for streaming but in the end, reverting the changes would not actually fix anything, because if your hard drive is slower or the preset is even more complex or the timing is randomly off you get the same issues again.

**Comment by klepp0906** [comment]
OBS is just used to stream to like, youtube and twitch.  AKA let other people watch you play.  

obs isnt a client that lets you play retroarch remotely.   Thats what i mean by streaming.   Remotely playing retroarch on my TV, not uploading live gaming footage to youtube etc.

**Comment by LazyBumHorse** [comment]
Ah, in-home streaming! Okay, got it.

**Comment by LazyBumHorse** [comment]
Same applied for "regular" streaming as well though

**Comment by klepp0906** [comment]
of a for the video shader setting,  I didnt realize it held onto the old shader and used it when u loaded something else.   Never noticed but it didnt seem to.

Aka you mean if I loaded something via a playlist, we'll say beetle hw w/ the crt/venom shader.  Closed it, then loaded something from a different playlist that didnt have an override, it would automatically load crt/venom?

So it was behaving like overlay settings still do (;p)  Id say that was certainly good to remove then.

**Comment by klepp0906** [comment]
> Same applies for "regular" streaming as well though

yea if this/that applies to regular streaming as well, i can definitely see this being a bigger issue.   Not that streaming retroarch is wildly popular, but some people do it for sure, and that opens up more people to accommodate with your fix, which makes me feel even less bad for being a bother ^^

**Comment by LazyBumHorse** [comment]
> Aka you mean if I loaded something via a playlist, we'll say beetle hw w/ the crt/venom shader. Closed it, then loaded something from a different playlist that didnt have an override, it would automatically load crt/venom?

Yes, _almost_!
It's actually related to the shader menu, each time you loaded a preset or applied one in there, it would save it via the `video_shader` setting.
And whenever you opened any content without any core/folder/game preset, it would actually load the shader from the `video_shader` setting.
This made it almost impossible to have games which _don't_ use shaders.

**Comment by klepp0906** [comment]
ah, i see.  again same problem with the onscreen overlay setting currently lol.   Have to create overrides in all situations pretty much once you start using them.

which is fine for me, but im really particular about having everything set up and complete as RA is the only frontend i use, and will ever use ;p  Can definitely see where its not ideal for the average joe though.

**Comment by LazyBumHorse** [comment]
If you got some time, you can test this one:
`git fetch origin`
`git checkout -B shader_delay origin/shader_delay`

It should reapply the loaded shader again after 3 seconds, hopefully restoring the image to what it should look like.

**Comment by klepp0906** [comment]
sorry, been outside prepping the yard for my kids bday party all day.   Let me hit a shower and ill build it and let'er rip

**Comment by LazyBumHorse** [comment]
No rush, take your time

**Comment by klepp0906** [comment]
same behavior.  frozen black screen with audio.

I dont think reapplying the shader would do anything as the display never shows up to begin with.  It freezes before anything is on screen (except the border/overlay which isnt affected by shader stuff).

only reason i know its a "freeze" persay is because the `override loaded` popup displays, as does the xinput controller popup.   They both normally fade in a second or two, but in this case - remain stuck on screen.  (not sure if i ever mentioned that before or if it has relevance)

**Comment by LazyBumHorse** [comment]
Oh, so it actually freezes the screen?
Well, it doesn't really matters if it's glitched or frozen graphics, I guess :smile: 

You can give it another shot, this version should definitely work, since it won't try to load a shader until the 3 seconds are over.

**Comment by klepp0906** [comment]
ok gonna build it now.  will report back shortly.

quick question while i have your attention (since i imagine your privvy to some internal RA stuffs) lol.

Totally off topic - but im working on setting up Philips CD-i and its a total pain.  Before I get invested I want to make sure I have the directories set up right for the future so to speak.  Do you know if retroarchs naming convention would be Philips - CD-i  or Philips - CDi?  or something else?

I dont want to get stuck renaming everything under the sun down the road if i can avoid it when/if it ever gets a proper database and thumbnail index etc lol

thread derail by the OP :p

**Comment by LazyBumHorse** [comment]
> Do you know if retroarchs naming convention would be Philips - CD-i or Philips - CDi? or something else?

Absolutely no idea, I'd guess "Philips - CD-i" because "CD-i" is the official name and there's "Philips - Videopac+" (whatever that is) on the [thumbnails server](http://thumbnails.libretro.com/).

**Comment by klepp0906** [comment]
yea thats what i was leaning towards, but in RetroArch\assets\xmb\automatic\png i found some files named Philips - CDi

Hrm.  Any idea who id ask?

As for the build.  It worked the first time, if i tried to load the game again, froze with an all white screeen now.

I tried it a second time to make sure (closed out stream entirely and restarted) and resulted in the same exact behavior.

wonder why it doesnt affect geoms shader.  Due to how simple it is relative to crt guest dr venom i guess.

**Comment by LazyBumHorse** [comment]
> Any idea who id ask?

No idea. I don't really get what's the big issue though, isn't this just about a folder name?
If you have to rename a bunch of files, you can use tools like Bulk Rename Utility or similar and if it's about file content, you can always do a 'replace all'.

> wonder why it doesnt affect geoms shader. Due to how simple it is relative to crt guest dr venom i guess.

crt-geom only has a single shader pass, while crt-guest-dr-venom has 10(!) passes, making it tons faster to load and compile.

> As for the build. It worked the first time, if i tried to load the game again, froze with an all white screeen now.

Yeah, that's "intended", well, it's odd that it's a white screen instead of a black screen, but as long as the first time works, that's all I need to know and I can add this as a workaround (hidden behind a setting) for now.

**Comment by klepp0906** [comment]
well - its not just a single folder.  its several, and if RA bakes it in default then things will stop working.  Or ill have duplicates that'd need cleaning at best.

Ultimately just trying to avoid future work.   Looks like redump uses CD-i and as you mentioned, thats the proper name - I have to assume that'll be whats used should it come to pass.

Anyhow, onto the build.   I have seen the white screen before.   Its the shape of what the screen "would" be.  Aka it has the curve and corners rounded etc.  but its all white instead of showing the game /shrug.  

its popped up randomly testing all these builds ;p  99% of the time its just been the black screen though.

I do notice you said you added a 3 second delay.   When it did work - i could see it load w/out the shader - then flicker into existence proper.

Any chance we can try a 2 second delay?  See if we can avoid that odd adjustement being seen?

**Comment by klepp0906** [comment]
the only other caveat to this approach is now the notification bar across the bottom pops up saying a shader has been applied, where normally it wouldnt.    Not sure if you can remove that for the auto/delay/load cases if thats what you go with.

(hope that makes sense)

**Comment by LazyBumHorse** [comment]
This was just a proof of concept, the delay will of course be tweakable (through the setting I mentioned) and the message popup will of course not be there as well.

**Comment by klepp0906** [comment]
aaaaah interesting.   sounds like your 1 step ahead :)

whelp, keep me posted, will give it a gander if you'd like once you get a closer-to-final version or whatnot.

**Comment by LazyBumHorse** [comment]
This might already be the final version, you can give it a shot:
`git fetch origin`
`git checkout -B shader_delay origin/shader_delay`

You need to set `video_shader_delay` in your retroarch.cfg, e.g.
`video_shader_delay = "500"`
for a 500ms delay.

Make sure you don't make the same mistake as me and accidentally put the setting twice in the config.
As it turns out, RetroArch swaps multiple settings each restart, so I didn't get how it only worked every other time :joy:

**Comment by LazyBumHorse** [comment]
Btw, it's been merged (and it has a menu entry under Video settings if you enable Show Advanced Settings under User Interface) but I think the buildbot currently doesn't work with x64.
However, if you compile yourself, you already have 1.7.8, there's nothing really special about the stable builds.
I think you can also close the issue now.
